pennylane/templates/subroutines/trotter.py

```diff
@@ -0,0 +1,216 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains templates for Suzuki-Trotter approximation based subroutines.
+"""
+import copy
+
+import pennylane as qml
+
+# from pennylane.math ...
+from pennylane import numpy as np
+from pennylane.operation import Operation
+from pennylane.ops import Sum
+
+
+def _scalar(order):
+    """Compute the scalar used in the recursive expression.
+
+    Args:
+        order (int): order of trotter product (assume order is an even integer > 2).
+
+    Returns:
+        float: scalar to be used in the recursive expression.
+    """
+    root = 1 / (order - 1)
+    return (4 - 4**root) ** -1
+
+
+@qml.QueuingManager.stop_recording()
+def _recursive_expression(x, order, ops):
+    """Generate a list of operations using the
+    recursive expression which defines the trotter product.
+
+    Args:
+        x (complex): the evolution 'time'
+        order (int): the order of the Trotter Expansion
+        ops (Iterable(~.Operators)): a list of terms in the Hamiltonian
+
+    Returns:
+        List: the approximation as product of exponentials of the hamiltonian terms
+    """
+    if order == 1:
+        return [qml.exp(op, x * 1j) for op in ops]
+
+    if order == 2:
+        return [qml.exp(op, x * 0.5j) for op in (ops + ops[::-1])]
+
+    scalar_1 = _scalar(order)
+    scalar_2 = 1 - 4 * scalar_1
+
+    ops_lst_1 = _recursive_expression(scalar_1 * x, order - 2, ops)
+    ops_lst_2 = _recursive_expression(scalar_2 * x, order - 2, ops)
+
+    return (2 * ops_lst_1) + ops_lst_2 + (2 * ops_lst_1)
+
+
+class TrotterProduct(Operation):
+    r"""An operation representing the Suzuki-Trotter product approximation for the complex matrix exponential
+    of a given hamiltonian.
+
+    The Suzuki-Trotter product formula provides a method to approximate the matrix exponential of hamiltonian
+    expressed as a linear combination of terms which in general do not commute. Consider the hamiltonian
+    :math:`H = \Sigma^{N}_{j=0} O_{j}`, the product formula is constructed using symmetrized products of the terms
+    in the hamiltonian. The symmetrized products of order :math: `m \in [1, 2, 4, ..., 2k] | k \in \mathbb{N}`
+    are given by:
+
+    .. math::
+
+        \begin{align}
+            S_{m=1}(t) &= \Pi_{j=0}^{N} \ exp(i t O_{j}) \\
+            S_{m=2}(t) &= \Pi_{j=0}^{N} \ exp(i \frac{t}{2} O_{j}) \cdot \Pi_{j=N}^{0} \ exp(i \frac{t}{2} O_{j}) \\
+            &\vdots
+            S_{m=2k}(t) &= S_{2k-2}(p_{2k}t)^{2} \cdot S_{2k-2}((1-4p_{2k})t) \cdot S_{2k-2}(p_{2k}t)^{2}
+        \end{align}
+
+    Where the coefficient is :math:`p_{2k} = \frac{1}{4 - \sqrt[2k - 1]{4}}`.
+
+    The :math:`2k`th order, :math:`n`-step Suzuki-Trotter approximation is then defined as:
+
+    .. math:: exp(iHt) \approx (S_{2k}(\frac{t}{n}))^{n}
+
+    Args:
+        hamiltonian (Union[~.Hamiltonian, ~.Sum]):
```

Comment:
"time" is missing here?

Reply:
In the PRD, time is complex but implementation seems to be a real number. For me it makes more sense this way but just in case I comment it.

Reply:
yes nice catch! I will add time. It can still be complex, but we are also manually applying a factor of `1j` so that everything is complex to begin with.

Reply:
Should `time` show up here?